### PR TITLE
Phasing

### PR DESCRIPTION
**Pull request by trevorgerhardt**
Addresses #318 for Convert to Frequency modifications. Shouldn't be too hard to phase Add Trip Modifications to a Frequency Entry, just need to figure out how to handle stop selection in that case.

### CODE REVIEW

**Comment by mattwigway** [comment]
@trevorgerhardt can't we just use the graphical stop selector? I'm thinking we want to do this at all times, frequently the stop names are empty or not meaningful.

**Comment by mattwigway** [comment]
i.e. the same one we use for phasing?

**Comment by mattwigway** [comment]
i.e. the same one we use for reroutes?

File: lib/actions/__tests__/index.js [hunk 1]
```diff
@@ -63,15 +63,11 @@ describe('actions > index', () => {
     })
   })
 
-  describe('> map state', () => {
```
Comment (trevorgerhardt) [inline lib/actions/__tests__/index.js]: Meant to just group the user action.

File: lib/utils/timetable.js [hunk 2]
```diff
@@ -1,6 +1,9 @@
+import format from 'date-fns/format'
```
Comment (trevorgerhardt) [inline lib/utils/timetable.js]: Ideally I'd like us to move away from `moment.js` and use `date-fns` as much as possible.

File: lib/components/modification/phase.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+          <Select
+            name='Phase from stop'
+            onChange={this._setPhaseFromStop}
+            options={selectedPhaseFromTimetableStops.map((stop) =>
+              ({value: stop.stop_id, label: stop.stop_name}))}
+            placeholder='Select stop to phase from'
+            value={phaseFromStop}
+            />
+        </FormGroup>}
+      <InputNumber
```
Comment (trevorgerhardt) [inline lib/components/modification/phase.js]: 👌

**Comment by trevorgerhardt** [comment]
@mattwigway 

1. Stop selection with the GUI wasn't as straightforward of an implementation due to the nature of the selection but it can be done. The feed I was using had stop names so it appeared to be an appropriate first pass.
2. Didn't even think that you could phase to another FE within the modification 👌 
3. If we can't phase to new stops then we can't phase Add Trips modifications since they're all new stops...right? Or is it ok to phase AT a new stop FROM an existing stop? Cause then we could do it. 
4. 👌

**Comment by mattwigway** [comment]
1. Okay, yeah, let's go ahead and leave the stop names for now
3. Sorry, I was unclear. It's totally possible to phase add trips modifications (even to entries in the same modification), but you have to do it at stops which are present in the GTFS (i.e. the blue ones on the map, that have a stop ID and by extension a stop name, so the stop names approach works for now). The blue stops on the map indicate that the added trip stops at an existing stop, the black stops indicate a brand new stop (which we can't use for phasing yet) and the gray stops autogenerated stops based on the auto-create stops distance (which we should never phase to, since they move around)

File: lib/utils/timetable.js (hunk 2 above)
Comment (mattwigway) [inline lib/utils/timetable.js]: :+1:

**Comment by trevorgerhardt** [comment]
@mattwigway latest changes should now work with Add Trips modifications (with a corresponding update to the phasing branch in Analysis Server)

**Comment by mattwigway** [comment]
We should have a warning when you phase from an entry that doesn't have the same headway as the entry you're phasing. It's not necessarily wrong (many agencies have all lines arrive at a central location at the top of an hour, and then more frequent lines also arrive during that hour), but it is often a mistake. For instance, if you phase one line that runs every 12 minutes and another that runs every 15, they will only line up once per hour—probably not what was intended.

**Comment by mattwigway** [comment]
Nice work!

The "phase from stop" seems to be missing on add trips modifications. I'm also thinking the phasing stuff takes up a lot of screen real estate, maybe we should have a checkbox that turns it all on and off?

We also need to modify the R5 exporters.

**Comment by trevorgerhardt** [comment]
`Phase from timetable` will only show up when a `Phase at stop` is selected. Similarly with `Phase from stop` to `Phase from timetable`. We could apply that to `Phase minutes` also so that only one field is showing until it is in use.

^ likewise I imagine `Phase from stop` is not appearing because there are no stops available to phase from. This could be a code error, I'll add a warning and investigate.

**Comment by mattwigway** [comment]
Yeah, let's hide phaseMinutes.

**Comment by mattwigway** [comment]
Still not seeing the phase from stop tho:

image

Also, it occurs to me that any entry that has "exact times" selected should not be available for phasing (either as a from or to entry).

**Comment by mattwigway** [comment]
Looking good, although it looks like it's not possible to select the first stop as the phase from stop?

**Comment by mattwigway** [comment]
A few more comments on the R5 exporter:

- The `phaseAtStop` and `phaseFromStop` need to be feed-ID-qualified (e.g. they need to say `MBTA:5410` rather than `5410`)
- the `id` field on a timetable needs to be exported as `entryId`

**Comment by mattwigway** [comment]
I fixed the `entryId` issue and removed the use of the spread operator, but I'm not sure what to do about the feed-qualified stop IDs. We can't do the same thing we do with the source trips, using the feed of the modification, because the stops we are phasing with may come from a different feed.

**Comment by mattwigway** [comment]
This is _so close_, thanks for all your work on this @trevorgerhardt! I'm seeing the `phaseFromStops` in adjust frequency modifications being qualified twice, e.g. I'm getting `MBTA:MBTA:64` as `phaseFromStop`.

**Comment by mattwigway** [comment]
With the latest changes, it looks like `phaseSeconds` is not being exported. Additionally it looks like the phase from timetables are being named incorrectly:

image

The timetable in question runs from 7:00 AM until 10:00 PM, not sure why it's saying 2am to 5pm.

File: lib/selectors/modifications.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+
+export default function selectModifications (state) {
+  return state.scenario.modifications
```
Comment (mattwigway) [inline lib/selectors/modifications.js]: What's the advantage to using a selector here?

File: lib/components/modification/adjust-dwell-time.js [hunk 5]
```diff
@@ -35,34 +38,43 @@ export default class AdjustDwellTime extends Component {
...
+            selectedStops={selectedStops}
             update={update}
             />}
 
-        <br />
-
         <div className='form-group'>
-          <label htmlFor='adjust-dwell-time-scale' className='radio-inline'>
-            <input id='adjust-dwell-time-scale' type='radio' value='scale' checked={modification.scale} onChange={this._setScale} />Scale existing dwell times by
+          <label htmlFor='adjust-dwell-time-scale'>
```
Comment (trevorgerhardt) [inline lib/components/modification/adjust-dwell-time.js]: Yup!

File: lib/components/modification/editor.js [hunk 6]
```diff
@@ -152,51 +116,70 @@ export default class ModificationEditor extends Component {
...
+  setVariant
+}) {
+  return <div>
+    <legend>Active in variants</legend>
+    <div className='form-inline'>
+      {allVariants.map((v, i) => <Checkbox
+        checked={activeVariants[i]}
+        label={i + 1}
+        key={`variant-${i}-modification-${modificationId}`}
+        onChange={(e) => setVariant(i, e.target.checked)}
```
Comment (trevorgerhardt) [inline lib/components/modification/editor.js]: Totally. An easy/clearer way of doing it wasn't immediately apparent

File: lib/selectors/modifications.js (hunk 4 above)
Comment (trevorgerhardt) [inline lib/selectors/modifications.js]: Standardization of how we "retrieve" it. Could change in the future, but the input and output will be the same.

**Comment by trevorgerhardt** [comment]
🎉🎊🎉

**Pull request merged**